### PR TITLE
Added support for arbitrary hash functions that implement the hash.Ha…

### PR DESCRIPTION
**Pull request by lucasvo**
…sh interface

### CODE REVIEW

File: proofs/tree.go [hunk 1]
```diff
@@ -419,52 +435,58 @@ func CalculateProofNodeList(node, leafCount uint64) (nodes []*HashNode, err erro
...
 		if len(hashes[i].Left) == 0 {
-			hash = HashTwoValues(hash, hashes[i].Right)
+			hash = HashTwoValues(hash, hashes[i].Right, hashFunc)
 		} else {
-			hash = HashTwoValues(hashes[i].Left, hash)
+			hash = HashTwoValues(hashes[i].Left, hash, hashFunc)
 		}
 	}
 
+	fmt.Println("hash1", hash)
```
Comment (mikiquantum) [inline proofs/tree.go]: Maybe we do not need this.

Comment (lucasvo) [reply proofs/tree.go]: Maybe? I think quite certainly we don't need it ;)

File: proofs/tree.go [hunk 2]
```diff
@@ -386,13 +398,17 @@ func getIndexOfString(slice []string, match string) (index int, err error) {
...
-func HashTwoValues(a, b []byte) (hash []byte) {
+func HashTwoValues(a []byte, b []byte, hashFunc hash.Hash) (hash []byte) {
 	data := make([]byte, 64)
 	copy(data[:32], a[:32])
 	copy(data[32:], b[:32])
-	h := blake2b.Sum256(data)
-	hash = h[:]
-	return
+
+	defer hashFunc.Reset()
```
Comment (mikiquantum) [inline proofs/tree.go]: As we use hashing in a couple of places, I would wrap this into its own function called `performHashing` or similar. So it is reusable.

File: proofs/tree.go [hunk 3]
```diff
@@ -419,49 +435,53 @@ func CalculateProofNodeList(node, leafCount uint64) (nodes []*HashNode, err erro
...
-func CalculateHashForProofField(proof *Proof) (hash []byte, err error) {
+func CalculateHashForProofField(proof *Proof, hashFunc hash.Hash) (hash []byte, err error) {
 	input, err := ConcatValues(proof.Property, proof.Value, proof.Salt)
 	if err != nil {
 		return []byte{}, err
 	}
 
-	h := blake2b.Sum256(input)
-	hash = h[:]
+	defer hashFunc.Reset()
```
Comment (mikiquantum) [inline proofs/tree.go]: `performedHashing` function Reused here.

**Pull request merged**